lib/core/src/chain_swap.rs

```diff
@@ -0,0 +1,382 @@
+            .map(|c| c.to_string())
+            .or(swap.user_lockup_tx_id);
+
+        Self::validate_state_transition(swap.state, to_state)?;
+        self.persister.try_handle_chain_swap_update(
+            swap_id,
+            to_state,
+            server_lockup_tx_id,
+            user_lockup_tx_id,
+            claim_tx_id,
+            refund_tx_id,
+        )?;
+        if let Some(payment_id) = payment_id {
+            let _ = self.subscription_notifier.send(payment_id);
+        }
+        Ok(())
+    }
+
+    async fn claim(&self, chain_swap: &ChainSwap) -> Result<(), PaymentError> {
+        debug!("Initiating claim for Chain Swap {}", &chain_swap.id);
+        let refund_address = self.onchain_wallet.next_unused_address().await?.to_string();
+        let claim_tx_id = self.swapper.claim_chain_swap(chain_swap, refund_address)?;
+
+        if chain_swap.payment_type == PaymentType::Receive {
+            // We insert a pseudo-claim-tx in case LWK fails to pick up the new mempool tx for a while
+            // This makes the tx known to the SDK (get_info, list_payments) instantly
+            self.persister.insert_or_update_payment(PaymentTxData {
+                tx_id: claim_tx_id.clone(),
+                timestamp: None,
+                amount_sat: chain_swap.receiver_amount_sat,
+                fees_sat: 0,
+                payment_type: PaymentType::Receive,
+                is_confirmed: false,
+            })?;
+        }
+
+        self.update_swap_info(
+            &chain_swap.id,
+            Complete,
+            None,
+            None,
+            Some(&claim_tx_id),
+            None,
+        )
+        .await?;
+        Ok(())
+    }
+
+    async fn refund(&self, swap: &ChainSwap) -> Result<String, PaymentError> {
+        let amount_sat = swap.receiver_amount_sat;
+        let output_address = self.onchain_wallet.next_unused_address().await?.to_string();
+
+        let fee = self
+            .onchain_wallet
+            .build_tx(None, &output_address, amount_sat)
+            .await?
+            .all_fees()
+            .values()
+            .sum();
+
+        let refund_res = self
+            .swapper
+            .refund_chain_swap_cooperative(swap, &output_address, fee);
+        match refund_res {
+            Ok(res) => Ok(res),
+            Err(e) => {
+                warn!("Cooperative refund failed: {:?}", e);
+                self.refund_non_cooperative(swap, fee).await
+            }
+        }
+    }
+
+    async fn refund_non_cooperative(
+        &self,
+        swap: &ChainSwap,
+        broadcast_fees_sat: u64,
+    ) -> Result<String, PaymentError> {
+        info!(
+            "Initiating non-cooperative refund for Chain Swap {}",
+            &swap.id
+        );
+
+        let current_height = match swap.payment_type {
+            PaymentType::Receive => self.bitcoin_chain_service.tip()?.height as u32,
+            PaymentType::Send => self.onchain_wallet.tip().await.height(),
```

Comment:
Why not using the liquid chain service here?
Also I am still confused about the receive type. Do we support Receiving via BTC address in this PR?

Reply:
When receiving the lockup side is BTC, so we have to check the locktime relative to the BTC chain

Reply:
A lot of the receive side is covered in this PR, just not the interface. There are some open questions regarding handling of the receive (BTC-LBTC) refund. It would be a refund of the lockup to the BTC sender. Maybe we can have a call on Monday to step through it?

Reply:
Sure. My question was mainly why for the BTC chain we get the tip from the bitcoin chain service (electrum that is always up to date) and for liquid chain we get it from the onchain wallet (that may not be in sync)  and not the liquid_chain_service.

Reply:
Yeah right,  good point 👍